lib/pdk/cli/exec.rb

```diff
@@ -71,6 +41,108 @@ def self.try_vendored_bin(vendored_bin_path, fallback)
           fallback
         end
       end
+
+      # TODO: decide how/when to connect stdin to child process for things like pry
+      # TODO: need a way to set callbacks on new stdout/stderr data
+      class Command
+        attr_reader :argv
+        attr_reader :context
+        attr_accessor :timeout
+
+        def initialize(*argv)
+          @argv = argv
+
+          @process = ChildProcess.build(*@argv)
+          @process.leader = true
+
+          @stdout = Tempfile.new('stdout').tap { |io| io.sync = true }
+          @stderr = Tempfile.new('stderr').tap { |io| io.sync = true }
+
+          @process.io.stdout = @stdout
+          @process.io.stderr = @stderr
+
+          # Default to running things in the system context.
+          @context = :system
+        end
+
+        def context=(new_context)
+          unless [:system, :module].include?(new_context)
+            raise ArgumentError, _("Expected execution context to be :system or :module but got '%{context}'") % { context: new_contenxt }
+          end
+
+          @context = new_context
+        end
+
+        def add_spinner(message, opts = {})
+          @success_message = opts.delete(:success)
+          @failure_message = opts.delete(:failure)
+
+          @spinner = TTY::Spinner.new(message, opts)
+        end
+
+        def execute!
+          # Start spinning if configured.
+          @spinner.auto_spin if @spinner
+
+          if context == :module
+            # FIXME: manage ENV more precisely as well.
+
+            Dir.chdir(PDK::Util.module_root) do
+              ::Bundler.with_clean_env do
+                run_process!
+              end
+            end
+          else
+            run_process!
+          end
+
+          # Stop spinning when done (if configured).
+          if @spinner
+            if @process.exit_code.zero? && @success_message
```

Comment:
The spinner when configured will always leave at a "mid-spin" state b/c the `@success_message` or `@failure_message` variables aren't set. Do we want to process the stdout and stderr from the process for this? Or just set the spinner success/failure state with `:success` or `:failure` like you did in the implementation in `PDK::Util::Bundler`?